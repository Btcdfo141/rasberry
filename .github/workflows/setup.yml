name: Development environment setup

# yamllint disable-line rule:truthy
on:
  workflow_dispatch:
  schedule:
    - cron: "0 4 * * *"
  push:
    branches:
      - dev
      - rc
    paths:
      - "requirements.txt"
      - "requirements_all.txt"

jobs:
  setup:
    name: Initialize wheels builder
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, macos-latest]
    runs-on: ${{ matrix.os }}
    steps:
      - name: Checkout the repository
        uses: actions/checkout@v3.2.0

      - name: Run setup
        run: script/setup
