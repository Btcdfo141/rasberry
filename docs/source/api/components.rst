:mod:`spencerassistant.components`
===============================

air\_quality
--------------------------------------------

.. automodule:: spencerassistant.components.air_quality
    :members:
    :undoc-members:
    :show-inheritance:

alarm\_control\_panel
--------------------------------------------

.. automodule:: spencerassistant.components.alarm_control_panel
    :members:
    :undoc-members:
    :show-inheritance:

binary\_sensor
--------------------------------------------

.. automodule:: spencerassistant.components.binary_sensor
    :members:
    :undoc-members:
    :show-inheritance:

camera
---------------------------

.. automodule:: spencerassistant.components.camera
    :members:
    :undoc-members:
    :show-inheritance:

calendar
---------------------------

.. automodule:: spencerassistant.components.calendar
    :members:
    :undoc-members:
    :show-inheritance:

climate
---------------------------

.. automodule:: spencerassistant.components.climate
    :members:
    :undoc-members:
    :show-inheritance:

conversation
---------------------------

.. automodule:: spencerassistant.components.conversation
    :members:
    :undoc-members:
    :show-inheritance:

cover
---------------------------

.. automodule:: spencerassistant.components.cover
    :members:
    :undoc-members:
    :show-inheritance:

device\_tracker
---------------------------

.. automodule:: spencerassistant.components.device_tracker
    :members:
    :undoc-members:
    :show-inheritance:

fan
---------------------------

.. automodule:: spencerassistant.components.fan
    :members:
    :undoc-members:
    :show-inheritance:

light
---------------------------

.. automodule:: spencerassistant.components.light
    :members:
    :undoc-members:
    :show-inheritance:

lock
---------------------------

.. automodule:: spencerassistant.components.lock
    :members:
    :undoc-members:
    :show-inheritance:

media\_player
---------------------------

.. automodule:: spencerassistant.components.media_player
    :members:
    :undoc-members:
    :show-inheritance:

notify
---------------------------

.. automodule:: spencerassistant.components.notify
    :members:
    :undoc-members:
    :show-inheritance:

remote
---------------------------

.. automodule:: spencerassistant.components.remote
    :members:
    :undoc-members:
    :show-inheritance:

switch
---------------------------

.. automodule:: spencerassistant.components.switch
    :members:
    :undoc-members:
    :show-inheritance:

sensor
-------------------------------------

.. automodule:: spencerassistant.components.sensor
    :members:
    :undoc-members:
    :show-inheritance:

vacuum
-------------------------------------

.. automodule:: spencerassistant.components.vacuum
    :members:
    :undoc-members:
    :show-inheritance:

water\_heater
-------------------------------------

.. automodule:: spencerassistant.components.water_heater
    :members:
    :undoc-members:
    :show-inheritance:

weather
---------------------------

.. automodule:: spencerassistant.components.weather
    :members:
    :undoc-members:
    :show-inheritance:

webhook
---------------------------

.. automodule:: spencerassistant.components.webhook
    :members:
    :undoc-members:
    :show-inheritance:
