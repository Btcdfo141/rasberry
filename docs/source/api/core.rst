.. _core_module:

:mod:`spencerassistant.core`
-------------------------

.. automodule:: spencerassistant.core
    :members: