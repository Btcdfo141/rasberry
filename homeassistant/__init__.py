"""Init file for Home Assistant."""


test2
