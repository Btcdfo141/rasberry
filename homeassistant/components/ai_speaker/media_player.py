"""Support to interact with a AIS Player."""
import logging
from typing import Optional

from aisapi.ws import AisWebService
import requests

from homeassistant.components.media_player import (
    SUPPORT_NEXT_TRACK,
    SUPPORT_PAUSE,
    SUPPORT_PLAY,
    SUPPORT_PLAY_MEDIA,
    SUPPORT_PREVIOUS_TRACK,
    SUPPORT_SEEK,
    SUPPORT_SELECT_SOUND_MODE,
    SUPPORT_SELECT_SOURCE,
    SUPPORT_SHUFFLE_SET,
    SUPPORT_STOP,
    MediaPlayerEntity,
)
from homeassistant.components.media_player.const import (
    SUPPORT_BROWSE_MEDIA,
    SUPPORT_VOLUME_MUTE,
    SUPPORT_VOLUME_SET,
    SUPPORT_VOLUME_STEP,
)
from homeassistant.const import STATE_IDLE, STATE_OFF, STATE_PAUSED, STATE_PLAYING
from homeassistant.helpers import aiohttp_client
import homeassistant.util.dt as dt_util

from .const import DOMAIN
from .media_browser import async_browse_media

_LOGGER = logging.getLogger(__name__)

SUPPORT_AIS = (
    SUPPORT_PAUSE
    | SUPPORT_PREVIOUS_TRACK
    | SUPPORT_NEXT_TRACK
    | SUPPORT_PLAY_MEDIA
    | SUPPORT_PLAY
    | SUPPORT_STOP
    | SUPPORT_VOLUME_SET
    | SUPPORT_VOLUME_MUTE
    | SUPPORT_VOLUME_STEP
    | SUPPORT_SEEK
    | SUPPORT_SELECT_SOURCE
    | SUPPORT_SELECT_SOUND_MODE
    | SUPPORT_SHUFFLE_SET
    | SUPPORT_BROWSE_MEDIA
)


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up the AIS Player."""

    ais_player = AisPlayerDevice(config_entry.data, hass)
    async_add_entities([ais_player], True)


class AisPlayerDevice(MediaPlayerEntity):
    """Representation of a AIS Player ."""

    def turn_off(self):
        """Turn off the player."""

    def clear_playlist(self):
        """Clear the media player playlist."""

    def set_repeat(self, repeat):
        """Set the repeat for the AIS player."""

    def turn_on(self):
        """Turn on the player."""

    def __init__(self, config_entry_data, hass):
        """Initialize the Ais Player device."""
        self._ais_info = config_entry_data.get("ais_info")
        self._ais_id = self._ais_info.get("ais_id")
        self._ais_ws_url = self._ais_info.get("ais_url")
        self._status = None
        self._playing = False
        self._stream_image = None
        self._media_title = None
        self._media_source = None
        self._album_name = None
        self._sound_mode = "NORMAL"
        self._media_status_received_time = None
        self._media_position = 0
        self._duration = 0
        self._media_id = None
        self._volume_level = 0.5
        self._shuffle = False
        self._web_session = aiohttp_client.async_get_clientsession(hass)
        self._ais_gate = AisWebService(hass.loop, self._web_session, self._ais_ws_url)

    @property
    def device_info(self):
        """Device info."""
        return {
            "identifiers": {(DOMAIN, self._ais_id)},
            "name": "AI-Speaker",
            "manufacturer": self._ais_info["Manufacturer"],
            "model": self._ais_info["Model"],
            "sw_version": self._ais_info["OsVersion"],
            "via_device": None,
        }

    def set_volume_level(self, volume):
        """Set the volume level."""
        self._volume_level = volume
        vol = int(volume * 100)
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "setVolume", "val": vol, "ais_url": self._ais_ws_url},
        )

    def select_sound_mode(self, sound_mode):
        """Set the sound mode."""
        self._sound_mode = sound_mode

    async def async_fetch_status(self):
        """Fetch status from AIS."""
        audio_status = await self._ais_gate.get_audio_status()
        if audio_status is not None:
            self._volume_level = audio_status.get("currentVolume", 100) / 100
            self._status = audio_status.get("currentStatus", 0)
            self._playing = audio_status.get("playing", False)
            self._media_position = audio_status.get("currentPosition", 0)
            self._duration = audio_status.get("duration", 0)
            self._stream_image = audio_status.get("media_stream_image", None)
            self._media_title = audio_status.get("currentMedia", "AI-Speaker")
            self._media_source = audio_status.get("media_source", self._media_source)
            self._album_name = audio_status.get("media_album_name", "AI-Speaker")
        else:
            _LOGGER.error("Fetch status from AIS")

    def select_source(self, source):
        """Choose a different available playlist and play it."""

    @property
    def volume_level(self):
        """Return the volume level of the Player."""
        return self._volume_level

    @property
    def media_duration(self):
        """Return the duration of current playing media in seconds."""
        return float(self._duration) / 1000

    @property
    def shuffle(self):
        """Boolean if shuffle is enabled."""
        return self._shuffle

    def set_shuffle(self, shuffle):
        """Enable/disable shuffle mode."""
        self._shuffle = shuffle
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {
                "key": "setPlayerShuffle",
                "val": self._shuffle,
                "ais_url": self._ais_ws_url,
            },
        )

    def media_seek(self, position):
        """Seek the media to a specific location."""
        if position == 0:
            val = -5000
        elif position == 1:
            val = 5000
        else:
            val = position * 1000
            self._media_status_received_time = dt_util.utcnow()
            self._media_position = position * 1000
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "seekTo", "val": val, "ais_url": self._ais_ws_url},
        )

    def volume_up(self):
        """Service to send the exo the command for volume up."""
        self._volume_level = min(self._volume_level + 0.0667, 1)
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "upVolume", "val": True, "ais_url": self._ais_ws_url},
        )

    def volume_down(self):
        """Service to send the exo the command for volume down."""
        self._volume_level = max(self._volume_level - 0.0667, 0)
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "downVolume", "val": True, "ais_url": self._ais_ws_url},
        )

    def mute_volume(self, mute):
        """Service to send the exo the command for mute."""
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "setVolume", "val": 0, "ais_url": self._ais_ws_url},
        )

    @property
    def sound_mode(self):
        """Return the current matched sound mode."""
        return self._sound_mode

    @property
    def sound_mode_list(self):
        """Return a list of available sound modes."""
        return [
            "NORMAL",
            "BOOST",
            "TREBLE",
            "POP",
            "ROCK",
            "CLASSIC",
            "JAZZ",
            "DANCE",
            "R&P",
        ]

    @property
    def available(self):
        """Return true if AIS Player is available and connected."""
        return True

    async def async_update(self):
        """Get the latest data and update the state."""
        await self.async_fetch_status()

    @property
    def name(self):
        """Return the name of the device."""
        return "AI-Speaker " + self._ais_info["Product"] + " player"

    @property
    def media_image_url(self):
        """Return the image url of current playing media."""
        return self._stream_image

    @property
    def state(self):
        """Return the media state."""
        if self._playing is False:
            return STATE_PAUSED
        if self._status == 1:
            return STATE_IDLE
        if self._status == 2:
            return STATE_PAUSED
        if self._status == 3:
            return STATE_PLAYING
        if self._status == 4:
            return STATE_PAUSED

        return STATE_OFF

    @property
    def media_title(self):
        """Return the title of current playing media."""
        return self._media_title

    @property
    def media_album_name(self):
        """Return album of current playing media."""
        return self._album_name

    @property
    def app_name(self):
        """Name of the current running app."""
        app = None
        if self._media_source is not None:
            app = self._media_source
            if self._album_name is not None:
                app = str(app) + " " + str(self._album_name)
        return app

    @property
    def supported_features(self):
        """Flag media player features that are supported."""
        return SUPPORT_AIS

    @property
    def media_content_id(self):
        """Return the media content id."""
        return self._media_id

    @property
    def media_stream_image(self):
        """Return the media content id."""
        return self._stream_image

    @property
    def media_position(self):
        """Position of current playing media in seconds."""
        position = float(self._media_position) // 1000
        return int(position)

    @property
    def device_state_attributes(self):
        """Return the specific state attributes of the player."""
        return self._ais_info

    @property
    def unique_id(self) -> Optional[str]:
        """Return a unique ID."""
        return self._ais_id

    @property
    def media_position_updated_at(self):
        """When was the position of the current playing media valid."""
        return self._media_status_received_time

    def media_play(self):
        """Service to send the AIS Player the command for play/pause."""
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "pauseAudio", "val": False, "ais_url": self._ais_ws_url},
        )
        self._playing = True
        self._status = 3

    def media_pause(self):
        """Service to send the AIS Player the command for play/pause."""
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "pauseAudio", "val": True, "ais_url": self._ais_ws_url},
        )
        self._playing = False

    def media_stop(self):
        """Service to send the AIS Player the command for stop."""
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "pauseAudio", "val": True, "ais_url": self._ais_ws_url},
        )
        self._playing = False

    def media_next_track(self):
        """Service to send the AIS Player the command for next track."""
        self.hass.services.call(
            "ais_cloud", "play_next", {"media_source": self._media_source}
        )

    def media_previous_track(self):
        """Service to send the AIS Player the command for previous track."""
        self.hass.services.call(
            "ais_cloud", "play_prev", {"media_source": self._media_source}
        )

    def play_media(self, media_type, media_id, **kwargs):
        """Send the media player the command for playing a media."""
        if media_id.startswith("ais_tunein"):
            url_to_call = media_id.split("/", 3)[3]
            try:
                response_text = requests.get(url_to_call, timeout=2).text
                response_text = response_text.split("\n")[0]
                if response_text.endswith(".pls"):
                    response_text = requests.get(response_text, timeout=2).text
                    media_id = response_text.split("\n")[1].replace("File1=", "")
                elif response_text.startswith("mms:"):
                    response_text = requests.get(
                        response_text.replace("mms:", "http:"), timeout=2
                    ).text
                    media_id = response_text.split("\n")[1].replace("Ref1=", "")
                else:
                    media_id = response_text
            except Exception as error:  # pylint: disable=broad-except
                _LOGGER.e("AIS play_media error: %s", error)

        self._media_id = media_id
        self._media_position = 0
        self._media_status_received_time = dt_util.utcnow()
        self.hass.services.call(
            DOMAIN,
            "publish_command",
            {"key": "playAudio", "val": media_id, "ais_url": self._ais_ws_url},
        )

    async def async_browse_media(self, media_content_type=None, media_content_id=None):
        """Implement the media browsing helper."""
        result = await async_browse_media(
            media_content_type,
            media_content_id,
            self._ais_gate,
        )
        return result
