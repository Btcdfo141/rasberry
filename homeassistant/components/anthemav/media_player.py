"""Support for Anthem Network Receivers and Processors."""
from __future__ import annotations

import logging

import voluptuous as vol

from homeassistant.components.media_player import (
    PLATFORM_SCHEMA,
    MediaPlayerEntity,
    MediaPlayerEntityFeature,
)
from homeassistant.const import (
    CONF_HOST,
    CONF_MAC,
    CONF_NAME,
    CONF_PORT,
    STATE_OFF,
    STATE_ON,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.exceptions import ConfigEntryNotReady
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.dispatcher import (
    async_dispatcher_connect,
    async_dispatcher_send,
)
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType

from .const import (
    ANTHEMAV_UDATE_SIGNAL,
    CONF_MODEL,
    DEFAULT_NAME,
    DEFAULT_PORT,
    DOMAIN,
    MANUFACTURER,
)

DEVICE_CLASS_RECEIVER = "receiver"

_LOGGER = logging.getLogger(__name__)

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_HOST): cv.string,
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
    }
)


async def async_setup_platform(
    hass: HomeAssistant,
    config: ConfigType,
    async_add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
    """Set up our socket to the AVR."""
    _LOGGER.warning(
        "AnthemAV configuration is deprecated and has been automatically imported. Please remove the integration from your configuration file"
    )
    await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": config_entries.SOURCE_IMPORT},
        data=config,
    )


async def async_setup_entry(hass, config_entry, async_add_devices):
    """Set up entry."""
    name = config_entry.data[CONF_NAME]
    macaddress = config_entry.data[CONF_MAC]
    model = config_entry.data[CONF_MODEL]

    avr = hass.data[DOMAIN][config_entry.entry_id]

    if avr is None:
        raise ConfigEntryNotReady

    device = AnthemAVR(avr, name, macaddress, model)

    _LOGGER.debug("dump_devicedata: %s", device.dump_avrdata)
    _LOGGER.debug("dump_conndata: %s", avr.dump_conndata)

    async_add_devices([device])


class AnthemAVR(MediaPlayerEntity):
    """Entity reading values from Anthem AVR protocol."""

    _attr_should_poll = False
    _attr_supported_features = (
        MediaPlayerEntityFeature.VOLUME_SET
        | MediaPlayerEntityFeature.VOLUME_MUTE
        | MediaPlayerEntityFeature.TURN_ON
        | MediaPlayerEntityFeature.TURN_OFF
        | MediaPlayerEntityFeature.SELECT_SOURCE
    )

    def __init__(self, avr, name, macaddress, model):
        """Initialize entity with transport."""
        super().__init__()
        self.avr = avr
        self._attr_name = name
        self._unique_id = macaddress
        self._model = model
        self._attr_device_info = {
            "identifiers": {(DOMAIN, self._unique_id)},
            "name": name,
            "manufacturer": MANUFACTURER,
            "model": model,
        }

    def _lookup(self, propname, dval=None):
        return getattr(self.avr.protocol, propname, dval)

    async def async_added_to_hass(self):
        """When entity is added to hass."""
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                f"{ANTHEMAV_UDATE_SIGNAL}_{self._name}",
                self.async_write_ha_state,
            )
        )

    @property
    def state(self):
        """Return state of power on/off."""
        pwrstate = self._lookup("power")

        if pwrstate is True:
            return STATE_ON
        if pwrstate is False:
            return STATE_OFF
        return None

    @property
    def extra_state_attributes(self):
        """Return entity specific state attributes."""
        return {"attenuation": self._lookup("attenuation", -90)}

    @property
    def is_volume_muted(self):
        """Return boolean reflecting mute state on device."""
        return self._lookup("mute", False)

    @property
    def volume_level(self):
        """Return volume level from 0 to 1."""
        return self._lookup("volume_as_percentage", 0.0)

    @property
    def media_title(self):
        """Return current input name (closest we have to media title)."""
        return self._lookup("input_name", "No Source")

    @property
    def app_name(self):
        """Return details about current video and audio stream."""
        return (
            f"{self._lookup('video_input_resolution_text', '')} "
            f"{self._lookup('audio_input_name', '')}"
        )

    @property
    def source(self):
        """Return currently selected input."""
        return self._lookup("input_name", "Unknown")

    @property
    def source_list(self):
        """Return all active, configured inputs."""
        return self._lookup("input_list", ["Unknown"])

    @property
    def sound_mode_list(self):
        """Return all available sound mode."""
        if self.state is STATE_OFF:
            return None
        return self._lookup("audio_listening_mode_list", None)

    @property
    def sound_mode(self):
        """Return currently selected sound mode."""
        return self._lookup("audio_listening_mode_text", "Unknown")

    async def async_select_source(self, source):
        """Change AVR to the designated source (by name)."""
        self._update_avr("input_name", source)

    async def async_turn_off(self):
        """Turn AVR power off."""
        self._update_avr("power", False)

    async def async_turn_on(self):
        """Turn AVR power on."""
        self._update_avr("power", True)

    async def async_set_volume_level(self, volume):
        """Set AVR volume (0 to 1)."""
        self._update_avr("volume_as_percentage", volume)

    async def async_volume_up(self):
        """Turn volume up for media player."""
        volume = self.volume_level
        if volume < 1:
            await self.async_set_volume_level(min(1, volume + 0.04))

    async def async_volume_down(self):
        """Turn volume down for media player."""
        volume = self.volume_level
        if volume > 0:
            await self.async_set_volume_level(max(0, volume - 0.04))

    async def async_mute_volume(self, mute):
        """Engage AVR mute."""
        self._update_avr("mute", mute)

    async def async_select_sound_mode(self, sound_mode):
        """Switch the sound mode of the entity."""
        self._update_avr("audio_listening_mode_text", sound_mode)

    def _update_avr(self, propname, value):
        """Update a property in the AVR."""
        _LOGGER.debug("Sending command to AVR: set %s to %s", propname, str(value))
        setattr(self.avr.protocol, propname, value)

    @property
    def dump_avrdata(self):
        """Return state of avr object for debugging forensics."""
        attrs = vars(self)
        items_string = ", ".join(f"{item}: {item}" for item in attrs.items())
        return f"dump_avrdata: {items_string}"
