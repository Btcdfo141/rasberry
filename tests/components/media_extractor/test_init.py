"""The tests for Media Extractor integration."""
from typing import Any
from unittest.mock import patch

import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.media_extractor import DOMAIN
from homeassistant.components.media_player import SERVICE_PLAY_MEDIA
from homeassistant.core import HomeAssistant, ServiceCall
from homeassistant.setup import async_setup_component

from tests.components.media_extractor import MockYoutubeDL


async def test_play_media_service_is_registered(hass: HomeAssistant) -> None:
    """Test play media service is registered."""
    await async_setup_component(hass, DOMAIN, {DOMAIN: {}})
    await hass.async_block_till_done()

    assert hass.services.has_service(DOMAIN, SERVICE_PLAY_MEDIA)


@pytest.mark.parametrize(
    "config_fixture", ["empty_media_extractor_config", "audio_media_extractor_config"]
)
@pytest.mark.parametrize(
    ("media_content_id", "media_content_type"),
    [("https://www.youtube.com/watch?v=dQw4w9WgXcQ", "VIDEO")],
)
async def test_play_media_service(
    hass: HomeAssistant,
    mock_youtube_dl: MockYoutubeDL,
    calls: list[ServiceCall],
    snapshot: SnapshotAssertion,
    request: pytest.FixtureRequest,
    config_fixture: str,
    media_content_id: str,
    media_content_type: str,
) -> None:
    """Test play media service is registered."""
    config: dict[str, Any] = request.getfixturevalue(config_fixture)
    await async_setup_component(hass, DOMAIN, config)
    await hass.async_block_till_done()

    await hass.services.async_call(
        DOMAIN,
        SERVICE_PLAY_MEDIA,
        {
            "entity_id": "media_player.bedroom",
            "media_content_type": media_content_type,
            "media_content_id": media_content_id,
        },
    )
    await hass.async_block_till_done()

    assert calls[0].data == snapshot


async def test_download_error(
    hass: HomeAssistant,
    empty_media_extractor_config: dict[str, Any],
    calls: list[ServiceCall],
) -> None:
    """Test handling DownloadError."""

    await async_setup_component(hass, DOMAIN, empty_media_extractor_config)
    await hass.async_block_till_done()
    with patch(
        "homeassistant.components.media_extractor.YoutubeDL", return_value="yes"
    ):
        # mock.return_value.extract_info.side_effect = DownloadError

        await hass.services.async_call(
            DOMAIN,
            SERVICE_PLAY_MEDIA,
            {
                "entity_id": "media_player.bedroom",
                "media_content_type": "VIDEO",
                "media_content_id": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
            },
        )

    assert len(calls) == 0
