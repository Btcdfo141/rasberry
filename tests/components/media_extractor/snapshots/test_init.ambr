# serializer version: 1
# name: test_play_media_service
  ReadOnlyDict({
    'entity_id': 'media_player.bedroom',
    'extra': dict({
    }),
    'media_content_id': 'https://manifest.googlevideo.com/api/manifest/hls_playlist/expire/1694794256/ei/sC0EZYCPHbuZx_AP3bGz0Ac/ip/84.31.234.146/id/750c38c3d5a05dc4/itag/616/source/youtube/requiressl/yes/ratebypass/yes/pfa/1/wft/1/sgovp/clen%3D99471214%3Bdur%3D212.040%3Bgir%3Dyes%3Bitag%3D356%3Blmt%3D1694043438471036/hls_chunk_host/rr2---sn-5hnekn7k.googlevideo.com/mh/7c/mm/31,29/mn/sn-5hnekn7k,sn-5hne6nzy/ms/au,rdu/mv/m/mvi/2/pl/14/initcwndbps/2267500/vprv/1/playlist_type/DVR/dover/13/txp/4532434/mt/1694772337/fvip/3/short_key/1/keepalive/yes/fexp/24007246,24362685/beids/24350018/sparams/expire,ei,ip,id,itag,source,requiressl,ratebypass,pfa,wft,sgovp,vprv,playlist_type/sig/AOq0QJ8wRgIhAIC0iobMnRschmQ3QaYsytXg9eg7l9B_-UNvMciis4bmAiEAg-3jr6SwOfAGCCU-JyTyxcXmraug-hPcjjJzm__43ug%3D/lsparams/hls_chunk_host,mh,mm,mn,ms,mv,mvi,pl,initcwndbps/lsig/AG3C_xAwRQIhAOlqbgmuueNhIuGENYKCsdwiNAUPheXw-RMUqsiaB7YuAiANN43FxJl14Ve_H_c9K-aDoXG4sI7PDCqKDhov6Qro_g%3D%3D/playlist/index.m3u8',
    'media_content_type': 'VIDEO',
  })
# ---
# name: test_play_media_service[https://www.youtube.com/watch?v=dQw4w9WgXcQ-VIDEO-audio_media_extractor_config-]
  ReadOnlyDict({
    'entity_id': 'media_player.bedroom',
    'extra': dict({
    }),
    'media_content_id': 'https://manifest.googlevideo.com/api/manifest/hls_playlist/expire/1694805268/ei/tFgEZcu0DoOD-gaqg47wBA/ip/45.93.75.130/id/750c38c3d5a05dc4/itag/616/source/youtube/requiressl/yes/ratebypass/yes/pfa/1/wft/1/sgovp/clen%3D99471214%3Bdur%3D212.040%3Bgir%3Dyes%3Bitag%3D356%3Blmt%3D1694043438471036/hls_chunk_host/rr3---sn-5hne6nzy.googlevideo.com/mh/7c/mm/31,29/mn/sn-5hne6nzy,sn-5hnekn7k/ms/au,rdu/mv/m/mvi/3/pl/22/initcwndbps/1957500/vprv/1/playlist_type/DVR/dover/13/txp/4532434/mt/1694783146/fvip/2/short_key/1/keepalive/yes/fexp/24007246/sparams/expire,ei,ip,id,itag,source,requiressl,ratebypass,pfa,wft,sgovp,vprv,playlist_type/sig/AOq0QJ8wRQIhALAASH0_ZDQQoMA82qWNCXSHPZ0bb9TQldIs7AAxktiiAiASA5bQy7IAa6NwdGIOpfye5OgcY_BNuo0WgSdh84tosw%3D%3D/lsparams/hls_chunk_host,mh,mm,mn,ms,mv,mvi,pl,initcwndbps/lsig/AG3C_xAwRgIhAIsDcLGH8KJpQpBgyJ5VWlDxfr75HyO8hMSVS9v7nRu4AiEA2xjtLZOzeNFoJlxwCsH3YqsUQt-BF_4gikhi_P4FbBc%3D/playlist/index.m3u8',
    'media_content_type': 'VIDEO',
  })
# ---
# name: test_play_media_service[https://www.youtube.com/watch?v=dQw4w9WgXcQ-VIDEO-audio_media_extractor_config]
  ReadOnlyDict({
    'entity_id': 'media_player.bedroom',
    'extra': dict({
    }),
    'media_content_id': 'https://manifest.googlevideo.com/api/manifest/hls_playlist/expire/1694805268/ei/tFgEZcu0DoOD-gaqg47wBA/ip/45.93.75.130/id/750c38c3d5a05dc4/itag/616/source/youtube/requiressl/yes/ratebypass/yes/pfa/1/wft/1/sgovp/clen%3D99471214%3Bdur%3D212.040%3Bgir%3Dyes%3Bitag%3D356%3Blmt%3D1694043438471036/hls_chunk_host/rr3---sn-5hne6nzy.googlevideo.com/mh/7c/mm/31,29/mn/sn-5hne6nzy,sn-5hnekn7k/ms/au,rdu/mv/m/mvi/3/pl/22/initcwndbps/1957500/vprv/1/playlist_type/DVR/dover/13/txp/4532434/mt/1694783146/fvip/2/short_key/1/keepalive/yes/fexp/24007246/sparams/expire,ei,ip,id,itag,source,requiressl,ratebypass,pfa,wft,sgovp,vprv,playlist_type/sig/AOq0QJ8wRQIhALAASH0_ZDQQoMA82qWNCXSHPZ0bb9TQldIs7AAxktiiAiASA5bQy7IAa6NwdGIOpfye5OgcY_BNuo0WgSdh84tosw%3D%3D/lsparams/hls_chunk_host,mh,mm,mn,ms,mv,mvi,pl,initcwndbps/lsig/AG3C_xAwRgIhAIsDcLGH8KJpQpBgyJ5VWlDxfr75HyO8hMSVS9v7nRu4AiEA2xjtLZOzeNFoJlxwCsH3YqsUQt-BF_4gikhi_P4FbBc%3D/playlist/index.m3u8',
    'media_content_type': 'VIDEO',
  })
# ---
# name: test_play_media_service[https://www.youtube.com/watch?v=dQw4w9WgXcQ-VIDEO-empty_media_extractor_config-]
  ReadOnlyDict({
    'entity_id': 'media_player.bedroom',
    'extra': dict({
    }),
    'media_content_id': 'https://manifest.googlevideo.com/api/manifest/hls_playlist/expire/1694805294/ei/zlgEZcCPFpqOx_APj42f2Ao/ip/45.93.75.130/id/750c38c3d5a05dc4/itag/616/source/youtube/requiressl/yes/ratebypass/yes/pfa/1/wft/1/sgovp/clen%3D99471214%3Bdur%3D212.040%3Bgir%3Dyes%3Bitag%3D356%3Blmt%3D1694043438471036/hls_chunk_host/rr3---sn-5hne6nzy.googlevideo.com/mh/7c/mm/31,26/mn/sn-5hne6nzy,sn-aigzrnld/ms/au,onr/mv/m/mvi/3/pl/22/initcwndbps/2095000/vprv/1/playlist_type/DVR/dover/13/txp/4532434/mt/1694783390/fvip/1/short_key/1/keepalive/yes/fexp/24007246,24362685/beids/24350017/sparams/expire,ei,ip,id,itag,source,requiressl,ratebypass,pfa,wft,sgovp,vprv,playlist_type/sig/AOq0QJ8wRgIhANCPwWNfq6wBp1Xo1L8bRJpDrzOyv7kfH_J65cZ_PRZLAiEAwo-0wQgeIjPe7OgyAAvMCx_A9wd1h8Qyh7VntKwGJUs%3D/lsparams/hls_chunk_host,mh,mm,mn,ms,mv,mvi,pl,initcwndbps/lsig/AG3C_xAwRQIgIqS9Ub_6L9ScKXr0T9bkeu6TZsEsyNApYfF_MqeukqECIQCMSeJ1sSEw5QGMgHAW8Fhsir4TYHEK5KVg-PzJbrT6hw%3D%3D/playlist/index.m3u8',
    'media_content_type': 'VIDEO',
  })
# ---
# name: test_play_media_service[https://www.youtube.com/watch?v=dQw4w9WgXcQ-VIDEO-empty_media_extractor_config]
  ReadOnlyDict({
    'entity_id': 'media_player.bedroom',
    'extra': dict({
    }),
    'media_content_id': 'https://manifest.googlevideo.com/api/manifest/hls_playlist/expire/1694805294/ei/zlgEZcCPFpqOx_APj42f2Ao/ip/45.93.75.130/id/750c38c3d5a05dc4/itag/616/source/youtube/requiressl/yes/ratebypass/yes/pfa/1/wft/1/sgovp/clen%3D99471214%3Bdur%3D212.040%3Bgir%3Dyes%3Bitag%3D356%3Blmt%3D1694043438471036/hls_chunk_host/rr3---sn-5hne6nzy.googlevideo.com/mh/7c/mm/31,26/mn/sn-5hne6nzy,sn-aigzrnld/ms/au,onr/mv/m/mvi/3/pl/22/initcwndbps/2095000/vprv/1/playlist_type/DVR/dover/13/txp/4532434/mt/1694783390/fvip/1/short_key/1/keepalive/yes/fexp/24007246,24362685/beids/24350017/sparams/expire,ei,ip,id,itag,source,requiressl,ratebypass,pfa,wft,sgovp,vprv,playlist_type/sig/AOq0QJ8wRgIhANCPwWNfq6wBp1Xo1L8bRJpDrzOyv7kfH_J65cZ_PRZLAiEAwo-0wQgeIjPe7OgyAAvMCx_A9wd1h8Qyh7VntKwGJUs%3D/lsparams/hls_chunk_host,mh,mm,mn,ms,mv,mvi,pl,initcwndbps/lsig/AG3C_xAwRQIgIqS9Ub_6L9ScKXr0T9bkeu6TZsEsyNApYfF_MqeukqECIQCMSeJ1sSEw5QGMgHAW8Fhsir4TYHEK5KVg-PzJbrT6hw%3D%3D/playlist/index.m3u8',
    'media_content_type': 'VIDEO',
  })
# ---
